Firestore/core/src/firebase/firestore/api/document_reference.mm

```diff
@@ -0,0 +1,227 @@
+  // for an example.
+  dispatch_semaphore_t registered = dispatch_semaphore_create(0);
+  __block id<FIRListenerRegistration> listener_registration;
+  FIRDocumentSnapshotBlock listener = ^(FIRDocumentSnapshot* snapshot,
+                                        NSError* error) {
+    if (error) {
+      completion(nil, error);
+      return;
+    }
+
+    // Remove query first before passing event to user to avoid user actions
+    // affecting the now stale query.
+    dispatch_semaphore_wait(registered, DISPATCH_TIME_FOREVER);
+    [listener_registration remove];
+
+    if (!snapshot.exists && snapshot.metadata.fromCache) {
+      // TODO(dimond): Reconsider how to raise missing documents when offline.
+      // If we're online and the document doesn't exist then we call the
+      // completion with a document with document.exists set to false. If we're
+      // offline however, we call the completion handler with an error. Two
+      // options: 1) Cache the negative response from the server so we can
+      // deliver that even when you're
+      //    offline.
+      // 2) Actually call the completion handler with an error if the document
+      // doesn't exist when
+      //    you are offline.
+      completion(nil, [NSError errorWithDomain:FIRFirestoreErrorDomain
+                                          code:FIRFirestoreErrorCodeUnavailable
+                                      userInfo:@{
+                                        NSLocalizedDescriptionKey :
+                                            @"Failed to get document because "
+                                            @"the client is offline.",
+                                      }]);
+    } else if (snapshot.exists && snapshot.metadata.fromCache &&
+               source == FIRFirestoreSourceServer) {
+      completion(
+          nil,
+          [NSError
+              errorWithDomain:FIRFirestoreErrorDomain
+                         code:FIRFirestoreErrorCodeUnavailable
+                     userInfo:@{
+                       NSLocalizedDescriptionKey :
+                           @"Failed to get document from server. (However, "
+                           @"this "
+                           @"document does exist in the local cache. Run again "
+                           @"without setting source to "
+                           @"FIRFirestoreSourceServer to "
+                           @"retrieve the cached document.)"
+                     }]);
+    } else {
+      completion(snapshot, nil);
+    }
+  };
+
+  listener_registration = AddSnapshotListener(listener, options);
+  dispatch_semaphore_signal(registered);
+}
+
+id<FIRListenerRegistration> DocumentReference::AddSnapshotListener(
+    FIRDocumentSnapshotBlock listener, FSTListenOptions* options) {
+  FIRFirestore* firestore = firestore_;
+  FSTQuery* query = [FSTQuery queryWithPath:key_.path()];
+  DocumentKey key = key_;
+
+  ViewSnapshotHandler snapshotHandler = [key, listener, firestore](
```

Comment:
Rename to `snapshot_handler` or just `handler` given the crazy indentation of the argument list. Can this be split near the `=` to make this read better?

Reply:
Done.